test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriBuilder/ContextUrlPathStringTests.cs

```diff
@@ -17,51 +18,259 @@ public class ContextUrlPathStringTests : UriBuilderTestBase
-            string res = this.GetContextUrlPathString(queryUri);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
             Assert.Equal("Dogs", res);
         }
 
         [Fact]
         public void ContextUrlPathWithNavigationPropertyLinks()
         {
             Uri queryUri = new Uri("People(1)/MyDog/$ref", UriKind.Relative);
-            string res = this.GetContextUrlPathString(queryUri);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
             Assert.Equal("People(1)", res);
         }
 
         [Fact]
         public void ContextUrlPathWithSimpleServiceOperation()
         {
             Uri queryUri = new Uri("GetCoolPeople", UriKind.Relative);
-            string res = this.GetContextUrlPathString(queryUri);
-            Assert.Equal("", res);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
+            Assert.Equal("People", res);
         }
 
         [Fact]
         public void ContextUrlPathWithComplexServiceOperationIsComposable()
         {
             Uri queryUri = new Uri("GetSomeAddress/City", UriKind.Relative);
-            string res = this.GetContextUrlPathString(queryUri);
-            Assert.Equal("City", res);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
+            Assert.Equal("Edm.String", res);
         }
 
         [Fact]
         public void ContextUrlPathWithEntityServiceOperationIsComposable()
         {
             Uri queryUri = new Uri("GetCoolestPerson/Fully.Qualified.Namespace.Employee", UriKind.Relative);
-            string res = this.GetContextUrlPathString(queryUri);
-            Assert.Equal("Fully.Qualified.Namespace.Employee", res);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
+            Assert.Equal("People/Fully.Qualified.Namespace.Employee", res);
         }
 
-        #region private methods
-        private string GetContextUrlPathString(Uri queryUri)
+        [Theory]
+        [InlineData("People(1)/Fully.Qualified.Namespace.GetSomeAddressFromPerson()/Street")]
+        [InlineData("People(1)/Fully.Qualified.Namespace.GetSomeAddressFromPerson/Street")]
+        public void ContextUrlPathWithComplexPropertyAccessAfterOperationIsComposable(string uri)
         {
-            ODataUriParser odataUriParser = new ODataUriParser(HardCodedTestModel.TestModel, queryUri);
+            Uri queryUri = new Uri(uri, UriKind.Relative);
+            string res = this.GetContextUrlPathString(HardCodedTestModel.TestModel, queryUri);
+            Assert.Equal("Edm.String", res);
+        }
+
+
+        [Theory]
+        [InlineData("Customers(1)/NS.GetSomeOrders")]
+        [InlineData("Customers(1)/NS.GetSomeOrders()")]
+        [InlineData("Customers(1)/NS.GetAnOrder")]
+        [InlineData("Customers(1)/NS.GetAnOrder()")]

```

Comment:
Do we have a test (somewhere) that, in the case of GetAnOrder, the contextUrl written has the /$entity suffix?  I believe this happens in ODataContextUriBuilder.CreateFromContextUrlInfo, but didn't verify all of the code paths.

Reply:
I added one extra test case to verify the context uri builder.

Reply:
Awesome; thanks!

---
In reply to: [159340489](https://github.com/OData/odata.net/pull/1023#discussion_r159340489) [](ancestors = 159340489)